include/swift/SIL/SILBuilder.h

```diff
@@ -31,27 +31,33 @@ class IntegerLiteralExpr;
 class FloatLiteralExpr;
 class SILGlobalVariable;
 
-class SILBuilder {
-  friend class SILBuilderWithScope;
-
-  SILFunction *F;
-  SILModule &Mod;
+/// Manage the state needed for a SIL pass across multiple, independent
+/// SILBuilder invocations.
+///
+/// A SIL pass can instantiate a SILBuilderContext object to track information
+/// across multiple, potentially independent invocations of SILBuilder. This
+/// allows utilities used within the pass to construct a new SILBuilder instance
+/// whenever it is convenient or appropriate. For example, a separate SILBuilder
+/// should be constructed whenever the current debug location or insertion point
+/// changed. Reusing the same SILBuilder and calling setInsertionPoint() easily
+/// leads to incorrect debug information.
+class SILBuilderContext {
+  friend class SILBuilder;
+
+  using InsertedInstVec = SmallVectorImpl<SILInstruction *>;
+
+  SILModule &Module;
 
   /// Allow the SIL module conventions to be overriden within the builder.
   /// This supports passes that lower SIL to a new stage.
-  SILModuleConventions silConv = SILModuleConventions(Mod);
-
-  /// If this is non-null, the instruction is inserted in the specified
-  /// basic block, at the specified InsertPt.  If null, created instructions
-  /// are not auto-inserted.
-  SILBasicBlock *BB;
-  SILBasicBlock::iterator InsertPt;
-  const SILDebugScope *CurDebugScope = nullptr;
-  Optional<SILLocation> CurDebugLocOverride = None;
+  SILModuleConventions silConv = SILModuleConventions(Module);
 
   /// If this pointer is non-null, then any inserted instruction is
   /// recorded in this list.
-  SmallVectorImpl<SILInstruction *> *InsertedInstrs = nullptr;
+  ///
+  /// TODO: Give this ownership of InsertedInstrs and migrate users that
+  /// currently provide their own InsertedInstrs.
+  InsertedInstVec *InsertedInstrs = nullptr;
```

Comment:
Two questions here:

1. Why use the typedef here and not everywhere else (i.e. constructor)? What value does it provide? I would rather have the real type here given it is just SmallVectorImpl<SILInstruction *>.
2. What are you imagining as the new design? The builder with a TinyPtrVector?

Reply:
It looks like I just forgot to remove the useless typedef. I would imagine a plain old SmallVector would be fine, and we could reexpose that to clients if needed as a SmallVectorImpl. InsertedInstrs could be moved into a subclass if we feel like it. I added SILBuilderContext::notifyInserted(SILInstruction*) to make the long term intention more clear.